File: src/main/kotlin/cn/cangnova/cangjie/psi/CjBlockExpression.java
/*
 * Copyright 2024 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */

package cn.cangnova.cangjie.psi;

import cn.cangnova.cangjie.lang.CangJieLanguage;
import cn.cangnova.cangjie.lexer.CjTokens;
import cn.cangnova.cangjie.psi.psiUtil.CjElementUtilsKt;
import cn.cangnova.cangjie.psi.psiUtil.CjPsiUtilKt;
import cn.cangnova.cangjie.psi.psiUtil.PsiElementKt;
import cn.cangnova.cangjie.utils.ReadOnly;
import com.intellij.lang.Language;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.PsiFile;
import com.intellij.psi.impl.source.tree.CompositeElement;
import com.intellij.psi.impl.source.tree.LazyParseablePsiElement;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.util.PsiUtilCore;
import com.intellij.util.IncorrectOperationException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

import static cn.cangnova.cangjie.CjNodeTypes.BLOCK;

public class CjBlockExpression extends LazyParseablePsiElement implements CjElement, CjExpression, CjStatementExpression {
    public CjBlockExpression(@NotNull IElementType type,  @Nullable CharSequence text) {
        super(type, text);
    }
    public CjBlockExpression(@Nullable CharSequence text) {
        super(BLOCK, text);
    }


    @NotNull
    @Override
    public Language getLanguage() {
        return CangJieLanguage.INSTANCE;
    }

    @Override
    public String toString() {
        return getNode().getElementType().toString();
    }


    @Override
    public PsiFile getContainingFile() {
        return super.getContainingFile();
    }

    @Override
    public <T extends PsiElement> T getPsi(@NotNull Class<T> clazz) {
        return super.getPsi(clazz);
    }

    @NotNull
    @Override
    public CjFile getContainingCjFile() {
        return PsiElementKt.getContainingCjFile(this);
    }

    @Override
    public <D> void acceptChildren(@NotNull CjVisitor<Void, D> visitor, D data) {
        CjPsiUtil.visitChildren(this, visitor, data);
    }

    @Override
    public <R, D> R accept(@NotNull CjVisitor<R, D> visitor, @Nullable D data) {
        return visitor.visitBlockExpression(this, data);
    }

    @Override
    @SuppressWarnings("unchecked")
    public final void accept(@NotNull PsiElementVisitor visitor) {
        if (visitor instanceof CjVisitor) {
            accept((CjVisitor) visitor, null);
        } else {
            visitor.visitElement(this);
        }
    }

    @Override
    public void delete() throws IncorrectOperationException {
        CjElementUtilsKt.deleteSemicolon(this);
        super.delete();
    }

    @Override

    public PsiElement @NotNull [] getChildren() {
        PsiElement psiChild = getFirstChild();

        List<PsiElement> result = null;
        while (psiChild != null) {
            if (psiChild.getNode() instanceof CompositeElement) {
                if (result == null) result = new ArrayList<>();
                result.add(psiChild);
            }
            psiChild = psiChild.getNextSibling();
        }
        return result == null ? PsiElement.EMPTY_ARRAY : PsiUtilCore.toPsiElementArray(result);
    }

    @NotNull
    @Override
    public CjElement getPsiOrParent() {
        return this;
    }

    @Override
    public PsiElement getParent() {
        PsiElement substitute = CjPsiUtilKt.getParentSubstitute(this);
        return substitute != null ? substitute : super.getParent();
    }

    @Nullable
    public CjExpression getFirstStatement() {
        return findChildByClass(CjExpression.class);
    }

    @Nullable
    public CjExpression getLastStatement() {
        List<CjExpression> statement = getStatements();

        if (!statement.isEmpty()) {

            return statement.get(statement.size() - 1);
        }

        return null;
    }

    @ReadOnly
    @NotNull
    public List<CjExpression> getStatements() {
        return Arrays.asList(findChildrenByClass(CjExpression.class));
    }

    /**
     * 没有return关键字的语句
     *
     * @return
     */
    @ReadOnly
    @NotNull
    public Set<CjExpression> getStatementsWithoutReturnKeyword() {
        CjReturnExpression[] returns = findChildrenByClass(CjReturnExpression.class);


        Set<CjExpression> result = new HashSet<>();

        for (CjReturnExpression statement : returns) {
            result.add(statement.getReturnedExpression());
        }

        CjExpression lastStatement = getLastStatement();

        if (lastStatement != null) {
            if( !(lastStatement instanceof  CjReturnExpression)  && !(lastStatement instanceof CjDeclaration)  ){
                result.add(lastStatement);

            }
        }
        return result;
    }

    @ReadOnly
    @NotNull
    public Set<CjExpression> getReturnStatements() {
        Set<CjExpression> returns = new HashSet<>(List.of(findChildrenByClass(CjReturnExpression.class)));

        CjExpression lastStatement = getLastStatement();
        if (lastStatement != null) {
            returns.add(lastStatement);
        }
        return returns;
    }


    @Nullable
    public TextRange getLastBracketRange() {
        PsiElement rBrace = getRBrace();
        return rBrace != null ? rBrace.getTextRange() : null;
    }

    @Nullable
    public PsiElement getRBrace() {
        return findPsiChildByType(CjTokens.RBRACE);
    }

    @Nullable
    public PsiElement getLBrace() {
        return findPsiChildByType(CjTokens.LBRACE);
    }
}

File: psi/src/main/kotlin/org/cangnova/cangjie/psi/CjBlockExpression.kt
/*
 * Copyright 2026 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */
package org.cangnova.cangjie.psi

import com.intellij.lang.Language
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.PsiFile
import com.intellij.psi.impl.source.tree.CompositeElement
import com.intellij.psi.impl.source.tree.LazyParseablePsiElement
import com.intellij.psi.tree.IElementType
import com.intellij.psi.util.PsiUtilCore
import com.intellij.util.IncorrectOperationException
import org.cangnova.cangjie.lang.CangJieLanguage
import org.cangnova.cangjie.lexer.CjTokens
import org.cangnova.cangjie.psi.psiUtil.deleteSemicolon
import org.cangnova.cangjie.psi.psiUtil.getContainingCjFile
import org.cangnova.cangjie.psi.psiUtil.parentSubstitute
import java.util.*

open class CjBlockExpression : LazyParseablePsiElement, CjElement, CjExpression, CjStatementExpression {
    constructor(type: IElementType, text: CharSequence?) : super(type, text)
    constructor(text: CharSequence?) : super(CjNodeTypes.BLOCK, text)

    override fun getLanguage(): Language {
        return CangJieLanguage
    }

    override fun toString(): String {
        return node.elementType.toString()
    }
    override fun getContainingCjFile(): CjFile {
        return getContainingCjFile(this)

    }
    override fun getContainingFile(): PsiFile {
        return super.getContainingFile()
    }

    override fun <T : PsiElement> getPsi(clazz: Class<T>): T {
        return super.getPsi(clazz)
    }

    override fun <D> acceptChildren(visitor: CjVisitor<Unit, D>, data: D) {
        CjPsiUtil.visitChildren(this, visitor, data)
    }

    override fun <R, D> accept(visitor: CjVisitor<R, D>, data: D): R? {
        return visitor.visitBlockExpression(this, data)
    }

    override fun accept(visitor: PsiElementVisitor) {
        if (visitor is CjVisitor<*, *>) {
            @Suppress("UNCHECKED_CAST")
            accept(visitor as CjVisitor<Any?, Any?>, null as Any?)
        } else {
            visitor.visitElement(this)
        }
    }

    @Throws(IncorrectOperationException::class)
    override fun delete() {
        this.deleteSemicolon()
        super.delete()
    }

    override fun getChildren(): Array<PsiElement> {
        var psiChild = firstChild

        var result: MutableList<PsiElement>? = null
        while (psiChild != null) {
            if (psiChild.node is CompositeElement) {
                if (result == null) result = ArrayList()
                result.add(psiChild)
            }
            psiChild = psiChild.nextSibling
        }
        return if (result == null) PsiElement.EMPTY_ARRAY else PsiUtilCore.toPsiElementArray(result)
    }

    override fun getPsiOrParent(): CjElement {
        return this
    }

    override fun getParent(): PsiElement {
        val substitute = this.parentSubstitute
        return substitute ?: super.getParent()
    }



    val firstStatement: CjExpression?
        get() = findChildByClass(CjExpression::class.java)

    val lastStatement: CjExpression?
        get() {
            val statement = statements

            if (!statement.isEmpty()) {
                return statement[statement.size - 1]
            }

            return null
        }

    open val statements: List<CjExpression>
        get() = Arrays.asList(*findChildrenByClass(CjExpression::class.java))

    val statementsWithoutReturnKeyword: Set<CjExpression>
        /**
         * 没有return关键字的语句
         *
         * @return
         */
        get() {
            val returns =
                findChildrenByClass(CjReturnExpression::class.java)

            val result: MutableSet<CjExpression> = HashSet()

            for (statement in returns) {
                statement.returnedExpression?.let { result.add(it) }
            }

            val lastStatement = lastStatement

            if (lastStatement != null) {
                if (lastStatement !is CjReturnExpression && lastStatement !is CjDeclaration) {
                    result.add(lastStatement)
                }
            }
            return result
        }

    val returnStatements: Set<CjExpression>
        get() {
            val returns: MutableSet<CjExpression> = HashSet(
                java.util.List.of(
                    *findChildrenByClass(
                        CjReturnExpression::class.java,
                    ),
                ),
            )

            val lastStatement = lastStatement
            if (lastStatement != null) {
                returns.add(lastStatement)
            }
            return returns
        }

    val lastBracketRange: TextRange?
        get() {
            val rBrace = rBrace
            return rBrace?.textRange
        }

    val rBrace: PsiElement?
        get() = findPsiChildByType(CjTokens.RBRACE)

    val lBrace: PsiElement?
        get() = findPsiChildByType(CjTokens.LBRACE)
}
